Add the validation filter to block_sub_field()

kienstra opened the pull request:
* Adds an integration test for the repeater, with 2 rows of all possible controls.
* Though it doesn't test default values yet.
* [Adds a filter](https://github.com/getblocklab/block-lab/pull/419/files#diff-018b57db3c29f17ff59f382617e39f14R75) similar to the one for `block_field()`
* That filter is how each control's `validate()` function is called when calling `block_sub_field()` or `block_field()`
* For example, for the Image control, [Image::validate()](https://github.com/getblocklab/block-lab/blob/60e63eb6549a7fa1693d64e619278ae422de4b4f/php/blocks/controls/class-image.php#L59) outputs a different value if `$echo` is true (the URL), versus when it's false (the ID).

Closes #418 
Partially fulfills #404 (Add an integration test for the repeater)

kienstra committed:
Add the validation filter to block_sub_field()

This is how each control's validate()
function is called.
For example, for the Image control,
Image::validate() outputs a different value
if $echo is true (the URL), vs when it's fale (the ID).

kienstra commented:
This fixed the issue locally, but I should test it more.

kienstra committed:
Add an integration test for the Repeater field

Though this doesn't test default values,
this tests a repeater with
every possible control as a sub_field.

kienstra commented on tests/php/integration/helpers/class-abstract-attribute.php:
```diff
@@ -0,0 +1,201 @@
+ *
+ * @package Block_Lab
+ */
+abstract class Abstract_Attribute extends \WP_UnitTestCase {
```
Most of this class is simply abstracted out of [test-template-output.php](https://github.com/getblocklab/block-lab/pull/419/files#diff-65847bbff742889cb73a7e018e683093L16).

kienstra committed:
Remove needless escaping in test assertions, fix typos

There was a typo in a DocBlock,
where 'name' was spelled 'nam e'.

kienstra commented:
**Request For Review**

Hi @lukecarbis,
Hope you're doing great, I kind of miss Australia 😄 

Could you please review this when you have a chance? Thanks, Luke!

lukecarbis commented on php/post-types/class-block-post.php:
```diff
@@ -72,6 +72,7 @@ public function register_hooks() {
 		add_action( 'wp_insert_post_data', array( $this, 'save_block' ), 10, 2 );
 		add_action( 'init', array( $this, 'register_controls' ) );
 		add_filter( 'block_lab_field_value', array( $this, 'get_field_value' ), 10, 3 );
+		add_filter( 'block_lab_sub_field_value', array( $this, 'get_field_value' ), 10, 3 );
```
Great pickup!

kienstra replied:
Thanks!

lukecarbis reviewed (approved):
Australia misses you too!

This PR looks great. Here's the template I used to test:

```
if ( block_rows( 'repeater' ) ) {
	while ( block_rows( 'repeater' ) ) {
		block_row( 'repeater' );
		?>
		<p><?php echo block_sub_value( 'background' ); ?></p>
		<p><?php block_sub_field( 'background' ); ?></p>
		<?php
	}
}
reset_block_rows( 'repeater' );
```

Here's the result of that on `develop` (notice an unexpected value from `block_sub_field`): 
<img width="693" alt="Screen Shot 2019-09-09 at 7 39 44 am" src="https://user-images.githubusercontent.com/1097667/64494856-50625780-d2d5-11e9-9844-b99f2cffb8e9.png">

Here's the result of the same template on this branch (with the correct value for `block_sub_field`):
<img width="690" alt="Screen Shot 2019-09-09 at 7 40 20 am" src="https://user-images.githubusercontent.com/1097667/64494862-6ff98000-d2d5-11e9-9c86-a9a80502ff52.png">

kienstra committed:
Add an integration test for default values

Test all 13 fields that can have defaults,
though this doesn't include the repeater,
as it can't have a default.
@todo: create a test file for the repeater field's defaults.

kienstra commented:
Hi @lukecarbis,
Haha, thanks.

Thanks a lot for testing this, also.

If it's alright, I just pushed https://github.com/getblocklab/block-lab/pull/419/commits/7e7af7d3064ce6bb9d927b7abc6a83899fc367a7, which adds an integration test for default values (but not yet for the repeater). It only touches files in `tests/php/`

lukecarbis commented:
7e7af7d looks good! Feel free to merge.

kienstra commented:
Thanks, @lukecarbis!
